aten/src/ATen/native/sparse/cuda/SoftMax.cu

```diff
@@ -0,0 +1,597 @@
+    PackedTensorAccessor<scalar_t, 2> out_values_accessor,
+    PackedTensorAccessor<scalar_t, 2> grad_values_accessor) {
+  int tid = threadIdx.x;
+  int blkid = blockIdx.x;
+  int blksz = blockDim.x;
+  int gridsz = gridDim.x;
+
+  int index = tid + blkid * blksz;
+  int step = blksz * gridsz;
+
+  while (index < size) {
+    int64_t offset = pool_offsets[index];
+    int64_t* pool_indices = sorted_pool_indices + offset;
+    int64_t pool_indices_size = pool_sizes[index];
+
+    scalar_t* tmp_row = exp_sums_rows + index * nvalues;
+
+    /* Compute tmp = - sum_j output_j * grad_j */
+    for (int64_t p = 0; p < pool_indices_size; p++) {
+      auto i = pool_indices[p];
+      auto out_values_row = out_values_accessor[i];
+      auto values_row = values_accessor[i];
+      auto j = lower_bound_values[i];
+      if (j < grad_nnz && (out_offsets[i] == grad_offsets[j])) {
+        auto grad_values_row = grad_values_accessor[j];
+        for (int64_t k = 0; k < nvalues; k++) {
+          if (LogSoftMax) {
+            tmp_row[k] -= grad_values_row[k];
+          } else {
+            tmp_row[k] -= out_values_row[k] * grad_values_row[k];
+          }
+        }
+      }
+    }
+    /* Compute grad_input = output * (grad + tmp)*/
+    for (int64_t p = 0; p < pool_indices_size; p++) {
+      auto i = pool_indices[p];
+      auto out_values_row = out_values_accessor[i];
+      auto values_row = values_accessor[i];
+      auto j = lower_bound_values[i];
+
+      if (j < grad_nnz && (out_offsets[i] == grad_offsets[j])) {
+        auto grad_values_row = grad_values_accessor[j];
+        for (int64_t k = 0; k < nvalues; k++) {
+          if (LogSoftMax) {
+            values_row[k] = grad_values_row[k] +
+                c10::cuda::compat::exp(out_values_row[k]) * tmp_row[k];
+          } else {
+            values_row[k] =
+                out_values_row[k] * (grad_values_row[k] + tmp_row[k]);
+          }
+        }
+      } else {
+        for (int64_t k = 0; k < nvalues; k++) {
+          if (LogSoftMax) {
+            values_row[k] =
+                c10::cuda::compat::exp(out_values_row[k]) * tmp_row[k];
+          } else {
+            values_row[k] = out_values_row[k] * (tmp_row[k]);
+          }
+        }
+      }
+    }
+    index += step;
+  }
+}
+
+template <class scalar_t, bool requireMxRows = true>
+std::tuple<
+    thrust::device_vector<int64_t>,
+    thrust::device_vector<int64_t>,
+    thrust::device_vector<int64_t>,
+    thrust::device_vector<scalar_t>>
+compute_pool_max(
+    const Tensor& indices,
+    const Tensor& values,
+    const IntArrayRef& sizes,
+    int64_t nvalues,
+    const int64_t dim) {
+  auto ndim = indices.size(0);
+  auto nnz = indices.size(1);
+  thrust::host_vector<int64_t> host_strides(ndim, 1);
+
+  if (ndim > 1) {
+    for (int64_t i = ndim - 2; i >= 0; i--) {
+      host_strides[i] = host_strides[i + 1] * (i + 1 == dim ? 1 : sizes[i + 1]);
+    }
+  }
+  thrust::device_vector<int64_t> strides = host_strides;
```

Comment:
thrust::device_vector would use thrust's default allocator (cudaMalloc/cudaFree), those calls are synchronizing and expensive.